test/e2e/crds/v2/route.go

```diff
@@ -1041,4 +1043,263 @@ spec:
+    - name: proxy-rewrite
+      enable: true
+      config:
+        host: %s
+`, routeName, upstreamName, host)
+			var route apiv2.ApisixRoute
+			applier.MustApplyAPIv2(
+				types.NamespacedName{Namespace: s.Namespace(), Name: routeName},
+				&route,
+				routeSpec,
+			)
+		}
+
+		verifyAccess := func() {
+			By("verify access to external service")
+			request := func() int {
+				return s.NewAPISIXClient().GET("/ip").
+					WithHost("httpbin.org").
+					Expect().Raw().StatusCode
+			}
+			Eventually(request).WithTimeout(30 * time.Second).ProbeEvery(2 * time.Second).
+				Should(Equal(http.StatusOK))
+		}
+
+		It("access third-party service directly", func() {
+			createApisixUpstream(apiv2.ExternalTypeDomain, "httpbin.org")
+			createApisixRoute()
+			verifyAccess()
+		})
+
+		It("access third-party service with host rewrite", func() {
+			createApisixUpstream(apiv2.ExternalTypeDomain, "httpbin.org")
+			createApisixRouteWithHostRewrite("httpbin.org")
+			verifyAccess()
+		})
+
+		It("access external domain via ExternalName service", func() {
+			createExternalService("httpbin.org")
+			createApisixUpstream(apiv2.ExternalTypeService, externalServiceName)
+			createApisixRoute()
+			verifyAccess()
+		})
+
+		It("access in-cluster service via ExternalName", func() {
+			By("create temporary httpbin service")
+
+			By("get FQDN of temporary service")
+			fqdn := fmt.Sprintf("%s.%s.svc.cluster.local", "httpbin-service-e2e-test", s.Namespace())
+
+			By("setup external service and route")
+			createExternalService(fqdn)
+			createApisixUpstream(apiv2.ExternalTypeService, externalServiceName)
+			createApisixRoute()
+			verifyAccess()
+		})
+
+		Context("complex scenarios", func() {
+			It("multiple external services in one upstream", func() {
+				By("create ApisixUpstream with multiple external nodes")
+				upstreamSpec := `
+apiVersion: apisix.apache.org/v2
+kind: ApisixUpstream
+metadata:
+  name: httpbin-upstream
+spec:
+  externalNodes:
+  - type: Domain
+    name: httpbin.org
+  - type: Domain
+    name: postman-echo.com
+`
+				var upstream apiv2.ApisixUpstream
+				applier.MustApplyAPIv2(
+					types.NamespacedName{Namespace: s.Namespace(), Name: upstreamName},
+					&upstream,
+					upstreamSpec,
+				)
+
+				createApisixRoute()
+
+				By("verify access to multiple services")
+				time.Sleep(7 * time.Second)
+				hasEtag := false   // postman-echo.com
+				hasNoEtag := false // httpbin.org
+				for range 20 {
+					headers := s.NewAPISIXClient().GET("/ip").
+						WithHeader("Host", "httpbin.org").
+						WithHeader("X-Foo", "bar").
+						Expect().
+						Headers().Raw()
+					if _, ok := headers["Etag"]; ok {
+						hasEtag = true
+					} else {
+						hasNoEtag = true
+					}
+					if hasEtag && hasNoEtag {
+						break
+					}
+				}
+				assert.True(GinkgoT(), hasEtag && hasNoEtag, "both httpbin and postman should be accessed at least once")
+			})
+
+			It("should be able to use backends and upstreams together", func() {
```

Comment:
https://github.com/apache/apisix-ingress-controller/blob/master/test/e2e/crds/v2/route.go#L516-L545

They are very similar, I suggest integrating them into one context to avoid redundant use cases.